pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
 		//issue body.
 		DisableStrictZoneCheck bool
 	}
+	// [ServiceOverride "1"]
+	//  Service = s3
+	//  Region = region1
+	//  URL = https://s3.foo.bar
+	//  SigningRegion = signing_region
+	//
+	//  [ServiceOverride "2"]
+	//     Service = ec2
+	//     Region = region2
+	//     URL = https://ec2.foo.bar
+	//     SigningRegion = signing_region
+	ServiceOverride map[string]*struct {
+		Service       string
+		Region        string
+		URL           string
+		SigningRegion string
+	}
+}
+
+func (cfg *CloudConfig) validateOverrides() error {
+	if len(cfg.ServiceOverride) == 0 {
+		return nil
+	}
+	set := make(map[string]bool)
+	for onum, ovrd := range cfg.ServiceOverride {
+		name := strings.TrimSpace(ovrd.Service)
```

Comment:
Is the TrimSpace necessary? Doesn't gcfg do this?

Reply:
Just added a test. No gcfg does not do this. That means I will have to add string trim to the comparisons when ranging over the map. Can I alter the map when validating? That seems preferable to trimming on every lookup.

Reply:
By my testing it looks like gcfg does trim extra space:

```go
package main

import (
	"fmt"

	"gopkg.in/gcfg.v1"
)

type CloudConfig struct {
	ServiceOverride map[string]*struct {
		Region        string
		URL           string
		SigningRegion string
		Service       string
		SigningMethod string
	}
}

func main() {
	cfgStr := `
[serviceoverride "ec2_us-west-2"]
URL=https://ec2.foo.bar
SigningRegion=custom-signing-region
Region=us-west-2
Service=ec2

[serviceoverride "ec2_us-east-1"]
URL = https://ec2.foo.bar
SigningRegion = custom-signing-region
Region = us-east-1
Service = ec2
`
	// change the CloudConfig type to have the "ServiceOverride" map
	cfg := &CloudConfig{}
	err := gcfg.ReadStringInto(cfg, cfgStr)
	if err != nil {
		panic(fmt.Errorf("Failed to parse gcfg data: %s", err))
	}
	for _, override := range cfg.ServiceOverride {
		fmt.Printf("Region: '%s'\n", override.Region)
		fmt.Printf("URL: '%s'\n", override.URL)
		fmt.Printf("SigningRegion: '%s'\n", override.SigningRegion)
		fmt.Printf("Service: '%s'\n", override.Service)
	}
}
```

spits out:

```
Region: 'us-west-2'
URL: 'https://ec2.foo.bar'
SigningRegion: 'custom-signing-region'
Service: 'ec2'
SigningMethod: ''
Region: 'us-east-1'
URL: 'https://ec2.foo.bar'
SigningRegion: 'custom-signing-region'
Service: 'ec2'
```

Reply:
to add spaces in INI, you have to quote the string

Reply:
`{
			"Active Overrides",
			strings.NewReader(`
                [Global]

               [ServiceOverride "1"]
                Service = "s3      "
                Region = sregion
                URL = https://s3.foo.bar
                SigningRegion = sregion
                SigningMethod = sign
                `),
			nil,
			false, true,
			[]ServiceDescriptor{{name: "s3", region: "sregion"}},
		},
This failed without trims`